File: source-connector/src/test/java/sourceconnector/config/PipelineConfigTest.java
package sourceconnector.config;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.boot.context.properties.bind.Binder;
import org.springframework.boot.context.properties.source.MapConfigurationPropertySource;
import sourceconnector.config.util.YamlTestUtils;
import sourceconnector.domain.log.Log;
import sourceconnector.domain.processor.BaseProcessor;
import sourceconnector.domain.processor.impl.EmptyFilterProcessor;
import sourceconnector.domain.processor.impl.TrimMapperProcessor;

import java.io.IOException;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class PipelineConfigTest {

  @DisplayName("Should get empty list when no processor")
  @Test
  void ByPassProcessorTest() throws IOException {
    // given
    Map<String, Object> map = YamlTestUtils.getStringObjectMap("""
      transform:
        pipeline:
      """);
    Binder binder = new Binder(new MapConfigurationPropertySource(map));
    PipelineConfig pipelineConfig = binder.bind("transform", PipelineConfig.class).get();

    // when
    List<BaseProcessor<Log>> processorList = pipelineConfig.toProcessors();
    // then
    assertThat(processorList).isEmpty();
  }

  @DisplayName("Should get two processors when pipeline consists of two processors")
  @Test
  void twoProcessorsTest() throws IOException {
    // given
    Map<String, Object> map = YamlTestUtils.getStringObjectMap("""
      transform:
        pipeline:
          - type: Trim
          - type: SkipBlank
      """);
    Binder binder = new Binder(new MapConfigurationPropertySource(map));
    PipelineConfig pipelineConfig = binder.bind("transform", PipelineConfig.class).get();

    // when
    List<BaseProcessor<Log>> processorList = pipelineConfig.toProcessors();
    // then
    assertThat(processorList)
      .hasExactlyElementsOfTypes(TrimMapperProcessor.class, EmptyFilterProcessor.class);
  }

  @DisplayName("Should get processor irrelevant to case")
  @Test
  void ignoreTypeCaseTest() throws IOException {
    // given
    Map<String, Object> map = YamlTestUtils.getStringObjectMap("""
      transform:
        pipeline:
          - type: tRIM
          - type: skipBlank
      """);
    Binder binder = new Binder(new MapConfigurationPropertySource(map));
    PipelineConfig pipelineConfig = binder.bind("transform", PipelineConfig.class).get();

    // when
    List<BaseProcessor<Log>> processorList = pipelineConfig.toProcessors();
    // then
    assertThat(processorList)
      .hasExactlyElementsOfTypes(TrimMapperProcessor.class, EmptyFilterProcessor.class);
  }

  @DisplayName("Should throw IllegalArgumentException when incorrect case")
  @Test
  void caseSensitiveTypeTest() throws IOException {
    // given
    Map<String, Object> map = YamlTestUtils.getStringObjectMap("""
      transform:
        pipeline:
          - type: NotExist
      """);
    Binder binder = new Binder(new MapConfigurationPropertySource(map));
    PipelineConfig pipelineConfig = binder.bind("transform", PipelineConfig.class).get();

    // when then
    assertThatThrownBy(pipelineConfig::toProcessors)
      .isInstanceOf(IllegalArgumentException.class)
      .hasMessage("Invalid processor type: NotExist");
  }
}

File: source-connector/src/test/kotlin/sourceconnector/config/PipelineConfigTest.kt
package sourceconnector.config

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.boot.context.properties.bind.Binder
import org.springframework.boot.context.properties.source.MapConfigurationPropertySource
import sourceconnector.config.util.YamlTestUtils.getStringObjectMap
import sourceconnector.domain.log.Log
import sourceconnector.domain.processor.BaseProcessor
import sourceconnector.domain.processor.impl.EmptyFilterProcessor
import sourceconnector.domain.processor.impl.TrimMapperProcessor
import java.io.IOException

internal class PipelineConfigTest {
  @DisplayName("Should get empty list when no processor")
  @Test
  @Throws(IOException::class)
  fun byPassProcessorTest() {
    // given
    val map: Map<String, Any> = getStringObjectMap(
      """
      processing:
        pipelines:
      """.trimIndent()
    )
    val binder = Binder(MapConfigurationPropertySource(map))
    val pipelineConfig = binder.bind<PipelineConfig>("processing", PipelineConfig::class.java).get()

    // when
    val processorList: List<BaseProcessor<Log>> = pipelineConfig.toProcessors()
    // then
    assertThat<BaseProcessor<Log>>(processorList).isEmpty()
  }

  @DisplayName("Should get two processors when pipeline consists of two processors")
  @Test
  @Throws(IOException::class)
  fun twoProcessorsTest() {
    // given
    val map: Map<String, Any> = getStringObjectMap(
      """
      processing:
        pipelines:
          - type: Trim
          - type: SkipBlank
      """.trimIndent()
    )
    val binder = Binder(MapConfigurationPropertySource(map))
    val pipelineConfig = binder.bind<PipelineConfig>("processing", PipelineConfig::class.java).get()

    // when
    val processorList: List<BaseProcessor<Log>> = pipelineConfig.toProcessors()
    // then
    assertThat<BaseProcessor<Log>>(processorList)
      .hasExactlyElementsOfTypes(
        TrimMapperProcessor::class.java,
        EmptyFilterProcessor::class.java
      )
  }

  @DisplayName("Should get processor irrelevant to case")
  @Test
  @Throws(IOException::class)
  fun ignoreTypeCaseTest() {
    // given
    val map: Map<String, Any> = getStringObjectMap(
      """
      processing:
        pipelines:
          - type: tRIM
          - type: skipBlank
      """.trimIndent()
    )
    val binder = Binder(MapConfigurationPropertySource(map))
    val pipelineConfig = binder.bind<PipelineConfig>("processing", PipelineConfig::class.java).get()

    // when
    val processorList: List<BaseProcessor<Log>> = pipelineConfig.toProcessors()
    // then
    assertThat<BaseProcessor<Log>>(processorList)
      .hasExactlyElementsOfTypes(
        TrimMapperProcessor::class.java,
        EmptyFilterProcessor::class.java
      )
  }

  @DisplayName("Should throw IllegalArgumentException when incorrect case")
  @Test
  @Throws(IOException::class)
  fun caseSensitiveTypeTest() {
    // given
    val map: Map<String, Any> = getStringObjectMap(
      """
      processing:
        pipelines:
          - type: NotExist
      """.trimIndent()
    )
    val binder = Binder(MapConfigurationPropertySource(map))
    val pipelineConfig = binder.bind<PipelineConfig>("processing", PipelineConfig::class.java).get()

    // when then
    assertThatThrownBy { pipelineConfig.toProcessors() }
      .isInstanceOf(IllegalArgumentException::class.java)
      .hasMessage("Invalid processor type: NotExist")
  }
}
